File: apps/seller-android/app/src/main/java/app/orderak/seller/core/money/Money.kt
package app.orderak.seller.core.money

import java.text.NumberFormat
import java.util.Locale
import java.text.ParsePosition
import java.util.Currency as JavaCurrency
import kotlin.math.roundToLong

/**
 * Money as an amount plus a currency — never a bare number.
 *
 * Implements ADR-009, and mirrors `services/backend/src/platform/money/money.ts`.
 * The two cannot share code across Kotlin and TypeScript, so the shared artefact
 * is the API contract and these two files are kept deliberately parallel.
 *
 * What this replaces: `formatEgp(piasters)` and `parseEgpToPiasters(text)`,
 * which divided and multiplied by a literal 100. That is correct in Egypt and
 * wrong by a factor of ten in Kuwait, Bahrain and Oman, whose minor unit is a
 * thousandth:
 *
 *   15000 minor units  =  150.00 EGP
 *   15000 minor units  =   15.000 KWD
 *
 * `Countries.kt` already offers all three of those markets at sign-in.
 *
 * WHY THERE IS NO EXPONENT TABLE
 *   `java.util.Currency.getDefaultFractionDigits()` carries ISO 4217 exponents
 *   and has been available since API 1. A hand-written table would be a second
 *   source of truth that can only drift, and it would drift silently: a wrong
 *   exponent renders a plausible number rather than throwing.
 */

/** Currencies the app can render. Enabling one for sale is a server decision. */
val SUPPORTED_CURRENCIES = listOf("EGP", "SAR", "AED", "QAR", "KWD", "BHD", "OMR")

/**
 * An amount in the currency's smallest unit, plus the currency.
 *
 * `amountMinor` alone is meaningless: the same 15000 is 150 pounds or 15 dinars
 * depending on the field beside it.
 */
data class Money(val amountMinor: Long, val currency: String) {
    init {
        require(currency.length == 3) { "Currency must be an ISO 4217 alpha-3 code, got '$currency'" }
    }
}

/**
 * Minor units per major unit, read from ISO 4217 rather than declared.
 *
 * Throws on an unknown currency instead of defaulting to 2, because a default
 * turns "this currency was never added" into "every amount in it is wrong by a
 * factor of ten" — the exact failure this file exists to prevent.
 */
fun exponentOf(currency: String): Int =
    JavaCurrency.getInstance(currency).defaultFractionDigits

/**
 * How many minor units make one major unit: 100 for EGP, 1000 for KWD.
 *
 * The integer counterpart of the private pow10 below. Anything comparing two
 * amounts rather than rendering one should use this and stay in minor units —
 * floating point introduces a tolerance, and a tolerance is a place for a
 * factor-of-ten error to hide.
 */
fun minorUnitsPerMajor(currency: String): Long = when (exponentOf(currency)) {
    0 -> 1L
    2 -> 100L
    3 -> 1000L
    else -> throw IllegalArgumentException("Unsupported ISO 4217 exponent for $currency")
}

/**
 * Read a decimal string as minor units of [currency], or null if it is not a
 * valid amount in it.
 *
 * Null rather than a best guess. "12.345" is not an amount in a two-decimal
 * currency, and rounding it to 12.34 would invent a number nobody wrote — which
 * matters here because the caller is deciding whether a receipt matches an
 * order, and a fabricated near-miss is worse than no match at all.
 */
fun parseMinorUnits(text: String, currency: String): Long? {
    val exponent = exponentOf(currency)
    val dot = text.indexOf('.')
    val whole = if (dot < 0) text else text.substring(0, dot)
    val fraction = if (dot < 0) "" else text.substring(dot + 1)
    if (fraction.length > exponent) return null
    val major = whole.toLongOrNull() ?: return null
    val minor = if (exponent == 0) 0L else fraction.padEnd(exponent, '0').toLongOrNull() ?: return null
    return try {
        Math.addExact(Math.multiplyExact(major, minorUnitsPerMajor(currency)), minor)
    } catch (_: ArithmeticException) {
        // A digit run long enough to overflow is not an amount anyone transferred.
        null
    }
}

/**
 * A plain decimal for an editable field: no grouping separators, no symbol, and
 * trailing zeros trimmed so a whole amount reads "150" rather than "150.00".
 *
 * Separate from [formatMoney] on purpose. That one formats for reading and
 * groups thousands; this one produces something a seller can keep typing into
 * and [parseMoney] can read back without the round trip changing the number.
 */
fun majorUnitsText(money: Money): String {
    val per = minorUnitsPerMajor(money.currency)
    val major = money.amountMinor / per
    val minor = money.amountMinor % per
    if (minor == 0L) return major.toString()
    val digits = exponentOf(money.currency)
    return "$major." + minor.toString().padStart(digits, '0').trimEnd('0')
}

/**
 * Format for display, without the currency symbol.
 *
 * The screens wrap this in `R.string.currency_*`, so the symbol comes from the
 * string resources. The number of decimal places still comes from the currency.
 *
 * [locale] decides the digit shape and the grouping separator, and defaults to
 * the ambient one so existing callers are unchanged. A Compose caller should
 * pass the composition locale instead: the app switches language in-process via
 * AppCompatDelegate, and reading the ambient default there risks money keeping
 * Latin digits on a screen whose dates have already turned Arabic-Indic.
 */
fun formatMoney(money: Money, locale: Locale = Locale.getDefault()): String {
    val exponent = exponentOf(money.currency)
    val nf = NumberFormat.getNumberInstance(locale)
    nf.minimumFractionDigits = 0
    nf.maximumFractionDigits = exponent
    return nf.format(money.amountMinor / pow10(exponent))
}

/** Convenience for the common case of formatting a raw amount in a known currency. */
fun formatAmount(
    amountMinor: Long,
    currency: String,
    locale: Locale = Locale.getDefault(),
): String = formatMoney(Money(amountMinor, currency), locale)

/**
 * Parse a user-entered major-unit amount into minor units.
 *
 * The locale handling below is unchanged from `parseEgpToPiasters` and is
 * deliberately fussy — it was written against real input from Arabic, French and
 * numeric keyboards. Only the scaling factor is different: it comes from the
 * currency instead of a literal 100.
 */
fun parseMoney(text: String, currency: String): Money? {
    val trimmed = text.trim()
    if (trimmed.isEmpty()) return null

    // Numeric keyboards often emit a comma decimal even when the active locale
    // is English. Treat a one/two-digit suffix as decimals; keep "1,000" as a
    // locale-formatted grouping case.
    val commaSuffix = trimmed.substringAfterLast(',', missingDelimiterValue = "")
    val keyboardDecimal = if (
        ',' in trimmed && '.' !in trimmed && commaSuffix.length in 1..2 && commaSuffix.all(Char::isDigit)
    ) {
        trimmed.replace(',', '.').toDoubleOrNull()
    } else {
        null
    }

    // Parse with the active app locale first so Arabic/French digits and decimal
    // separators work. Require the entire input to be consumed.
    val position = ParsePosition(0)
    val localized = NumberFormat.getNumberInstance().parse(trimmed, position)
        ?.takeIf { position.index == trimmed.length }
        ?.toDouble()

    // Also accept a canonical dot decimal, which is common on numeric keyboards.
    val v = keyboardDecimal ?: localized ?: trimmed.toDoubleOrNull() ?: return null
    if (v < 0 || v.isNaN() || v.isInfinite()) return null
    // Round, don't floor — 4.35 must be 435 minor units, not 434. The server's
    // parseMoney() carries the same fix; this one found it first.
    return Money((v * pow10(exponentOf(currency))).roundToLong(), currency)
}

/**
 * Add amounts, refusing to mix currencies rather than producing a wrong total.
 */
operator fun Money.plus(other: Money): Money {
    require(currency == other.currency) { "Cannot add $currency to ${other.currency}" }
    return Money(amountMinor + other.amountMinor, currency)
}

/** Multiply by a whole quantity, as in a line item. */
operator fun Money.times(quantity: Int): Money = Money(amountMinor * quantity, currency)

/**
 * `Math.pow` on an Int exponent, kept exact.
 *
 * `10.0.pow(3)` is fine for the exponents ISO 4217 actually uses (0, 2 and 3),
 * but a table makes the intent obvious and removes any floating-point question
 * from a money path.
 */
private fun pow10(exponent: Int): Double = when (exponent) {
    0 -> 1.0
    2 -> 100.0
    3 -> 1000.0
    else -> throw IllegalArgumentException("Unsupported ISO 4217 exponent $exponent")
}

/** The app's default until a second market opens; see ADR-009. */
const val DEFAULT_CURRENCY = "EGP"

File: services/backend/src/platform/money/money.ts
/**
 * Money as an amount plus a currency — never a bare number.
 *
 * Implements ADR-009. The decision it replaces (ADR-002) stored money as
 * "integer piasters", which fused three separate facts into one: the amount,
 * the currency, and the number of minor units per major unit. That works
 * exactly as long as every market has 100 minor units. Three of the nine
 * markets `Countries.kt` already offers do not:
 *
 *   15000 minor units  =  150.00 EGP   (exponent 2)
 *   15000 minor units  =   15.000 KWD  (exponent 3)
 *
 * A hardcoded `/ 100` is not off by a rounding error in Kuwait, Bahrain and
 * Oman. It is off by a factor of ten.
 *
 * WHY THERE IS NO EXPONENT TABLE HERE
 *   `Intl.NumberFormat` already carries ISO 4217 exponents and is kept current
 *   by the runtime's ICU data. A hand-maintained table would be a second source
 *   of truth that can only ever drift away from the first, and it would drift
 *   silently — a wrong exponent produces a plausible number, not an error.
 *
 *   What is declared explicitly below is which currencies are *enabled*, which
 *   is a business decision and genuinely does belong in code.
 */

// `z` comes from @hono/zod-openapi rather than zod directly: it is the same
// Zod, already extended with `.openapi()`. Importing the bare zod here and
// calling `.openapi()` type-checks against the extended global declaration but
// throws at runtime, which is the worst combination of the two.
import { z } from "@hono/zod-openapi";

/**
 * Currencies this deployment accepts.
 *
 * ADR-009 changes the representation, not the market list: EGP is the only
 * enabled currency until the second market opens. The rest are declared so the
 * type, the validation and the tests all exercise a 3-exponent currency from
 * the first day rather than discovering it at launch.
 */
export const SUPPORTED_CURRENCIES = ["EGP", "SAR", "AED", "QAR", "KWD", "BHD", "OMR"] as const;
export const ENABLED_CURRENCIES = ["EGP"] as const;

/**
 * The currency assumed when a stored row or an inbound payload does not carry
 * one.
 *
 * Declared here so the assumption has exactly one home. It used to be a bare
 * `?? "EGP"` repeated across api-store.ts, catalog.ts and public-router.ts —
 * and in all three the SELECT feeding it never read the currency column at all,
 * so the fallback was not a fallback, it was the only value any of them ever
 * produced. A row stored as KWD came back out of the API as EGP.
 *
 * Opening a second market means changing the enabled list above, at which point
 * every remaining use of this constant is a place that has to be revisited
 * deliberately rather than one that quietly keeps answering "EGP".
 */
export const DEFAULT_CURRENCY: (typeof ENABLED_CURRENCIES)[number] = ENABLED_CURRENCIES[0];

export type Currency = (typeof SUPPORTED_CURRENCIES)[number];

export const CurrencySchema = z.enum(SUPPORTED_CURRENCIES);

/**
 * A monetary value. `amount_minor` is an integer count of the currency's
 * smallest unit — piasters for EGP, fils for KWD — and is meaningless without
 * the currency beside it.
 */
export const MoneySchema = z
	.object({
		amount_minor: z.number().int(),
		currency: CurrencySchema,
	})
	.openapi("Money", {
		description:
			"An amount in the currency's minor units, plus the currency. The number of minor "
			+ "units per major unit follows ISO 4217 and is not always 100: EGP has 2 decimal "
			+ "places, KWD has 3. Never divide by a constant.",
		example: { amount_minor: 15000, currency: "EGP" },
	});

export type Money = z.infer<typeof MoneySchema>;

/**
 * Minor units per major unit for a currency, read from ICU rather than declared.
 *
 * Throws on an unknown currency instead of defaulting to 2. A default here
 * would turn "we forgot to add this currency" into "every amount in this
 * currency is wrong by a factor of ten", which is the failure this module
 * exists to prevent.
 */
export function exponentOf(currency: Currency): number {
	const resolved = new Intl.NumberFormat("en", { style: "currency", currency }).resolvedOptions();
	const exponent = resolved.minimumFractionDigits;
	if (typeof exponent !== "number") {
		throw new Error(`No ISO 4217 exponent available for currency ${currency}`);
	}
	return exponent;
}

/** Build a Money, rejecting a non-integer amount rather than silently truncating it. */
export function money(amountMinor: number, currency: Currency): Money {
	if (!Number.isInteger(amountMinor)) {
		throw new Error(`Money amount must be an integer count of minor units, received ${amountMinor}`);
	}
	return { amount_minor: amountMinor, currency };
}

/**
 * Format for display. The locale governs digits and separators; the currency
 * governs the symbol and the number of decimal places.
 *
 * `ar-EG` renders Eastern Arabic numerals and `ar-SA` renders the Saudi riyal
 * symbol, both without any per-market branching here.
 */
export function formatMoney(value: Money, locale: string): string {
	const major = value.amount_minor / 10 ** exponentOf(value.currency);
	return new Intl.NumberFormat(locale, { style: "currency", currency: value.currency }).format(major);
}

/**
 * Parse a user-entered major-unit amount into minor units.
 *
 * Rounds rather than truncates: `4.35` in a 2-exponent currency is 435 minor
 * units, not 434. This repeats a fix already recorded in `Money.kt` — the
 * Android side hit the truncation bug first, and re-introducing it on the
 * server would be a regression against a known defect.
 *
 * Returns null for input that is not a finite number, so a caller cannot
 * mistake a parse failure for a zero amount.
 */
export function parseMoney(text: string, currency: Currency): Money | null {
	const normalized = text.replace(/[\s,٬ ]/g, "").replace(/٫/, ".");
	if (normalized === "" || !/^-?\d*\.?\d*$/.test(normalized)) return null;
	const major = Number(normalized);
	if (!Number.isFinite(major)) return null;
	return money(Math.round(major * 10 ** exponentOf(currency)), currency);
}

/** Add amounts, refusing to mix currencies rather than producing a wrong total. */
export function addMoney(left: Money, right: Money): Money {
	if (left.currency !== right.currency) {
		throw new Error(`Cannot add ${left.currency} to ${right.currency}`);
	}
	return money(left.amount_minor + right.amount_minor, left.currency);
}

/**
 * Apply a percentage, rounding half away from zero.
 *
 * The rounding mode is stated here because ADR-009 leaves it to the
 * implementation and a percentage of a 3-exponent amount is where it starts to
 * matter. Half-away-from-zero matches what `Math.round` does for positive
 * amounts and keeps refunds symmetric with charges.
 */
export function percentOf(value: Money, percent: number): Money {
	const raw = (value.amount_minor * percent) / 100;
	const rounded = raw < 0 ? -Math.round(-raw) : Math.round(raw);
	return money(rounded, value.currency);
}
